aptos-move/framework/aptos-framework/sources/Token.move

```diff
@@ -29,28 +40,99 @@ module AptosFramework::Token {
     // Core data structures for holding tokens
     //
 
-    /// Represents ownership of a the data associated with this Token
     struct Token has store {
         id: TokenId,
+        // the amount of tokens
         value: u64,
     }
 
+    /// global unique identifier of a token
+    struct TokenId has store, copy, drop {
+        // the id to the common token data shared by token with different serial number
+        token_data_id: TokenDataId,
+        // the serial number, every mint of this token_data_id will generate a new serial number
+        serial_number: u64
+    }
 
-    /// Represents a unique identity for the token
-    struct TokenId has copy, drop, store {
+    /// global unique identifier of tokendata
+    struct TokenDataId has copy, drop, store {
         // The creator of this token
         creator: address,
         // The collection or set of related tokens within the creator's account
-        collection: ASCII::String,
-        // Unique name within a collection within the creator's account
-        name: ASCII::String,
+        collection: String,
+        // the name of this token
+        name: String,
+    }
+
+    /// The royalty of a token
+    struct Royalty has copy, drop, store {
+        royalty_points_nominator: u64,
+        royalty_points_denominator: u64,
+        // if the token is jointly owned by multiple creators,
+        // the group of creators should create a shared account.
+        // the payee_address will be the shared account address.
+        payee_address: address,
+    }
+
+    /// This config specifies which fields in the Collection are mutable
+    struct CollectionMutabilityConfig has copy, store, drop {
+        // control if description is mutable
+        description: bool,
+        // control if uri is mutable
+        uri: bool,
+        // control if collection maxium is mutable
+        maximum: bool
+    }
+
+    /// This config specifies which fields in the TokenData are mutable
+    struct TokenMutabilityConfig has copy, store, drop {
+        // control if the token maximum is mutable
+        maximum: bool,
+        // control if the token uri is mutable
+        uri: bool,
+        // control if the token royalty is mutable
+        royalty: bool,
+        // control if the token description is mutable
+        description: bool,
+        // control if the default token property list are mutable
+        default_property_list: bool,
+        // control if token with different serial number can have different value
+        property_values: bool
+    }
+
+    /// The shared TokenData by tokens with different serial_number
+    struct TokenData has store {
+        // id of this token data
+        id: TokenDataId,
+        // the maxium of tokens can be minted from this token
+        maximum: u64,
```

Comment:
I think this used to be optional, I would leave it that way, there's no point in enforcing a maximum and requiring the user to track this or set an absurdly high number here to avoid transaction failures for creating too many nfts.

Reply:
We can do it in the SDK to provide users ergonomic APIs. It is much easier to do in SDK. One API could be the SDK just use U64_MAX or U128_MAX as maximum and don't ask users to provide the maximum.  Doing these ergonomic improvements in both smart contract and SDK seems a redundant of work to me and make the smart contract code complex, which is already too long.

We may argue there could be token that has more than U64_MAX. It is very rare and we can alway use a large U256 if necessary.